### PR TITLE
Add mingw

### PR DESCRIPTION
**Pull request by jkvatne**
I have now succeeded in compiling OpENer with mingw. The extra code is contained in MINGW subdirectories, besides the WIN32 and POSIX ones. Also the cmakeList.txt main file has added options to make for mingw. One source file had to be changed - the doublylinkedlist.c file had to include stdio.h in order to avoid missing declaration of NULL when tracing was off.

Quite a lot of warnings was initialy produced. The cmake file now adds options to turn them off, but the root problems should be fixed. 

Jan Kåre Vtne

### CODE REVIEW

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -72,6 +72,9 @@ endif( OpENer_TESTS )
 # OpENer C flags                      #
 #######################################
 set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall" )
+# The following extra flags are needed to compile without warnings on mingw
+# TODO: Fix the root problems and remove this line
+set( CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-unused-variable -Wno-incompatible-pointer-types -Wno-pointer-sign -Wno-format -Wno-discarded-qualifiers -Wno-unknown-pragmas" )
```
Comment (CapXilinx) [inline source/CMakeLists.txt]: Please do not disable warning, especially not for all platforms. Also, please do not put TODOs into source files, I currently try to get rid of them. Either this is an issue or it is not

**Comment by CapXilinx** [comment]
Hi,
thanks for the patch.
Could you please remove all the Makefiles and source/src/ports/MINGW/sample_application/cmake_install.cmake.
Could you also please rebase your commit to the latest develop commit?

**Comment by jkvatne** [comment]
This is too complicated for me. I give up.

**Comment by CapXilinx** [comment]
Hi Jan,

yes, I am currently trying this myself and it is super complicated :)
Don't worry I will fix it this week. Thanks for your contribution and time again!

Best regards,
Martin

**Comment by jkvatne** [comment]
Some things on my wish list: Get rid of all the warnings, compile it on Visual Studio (there are errors) and on mingw as part of CI. Also add the directory .idea to the .gitignore, so my Clion can keep its settings.  The generation of device.h is also quite difficult to get right, and could be skipped. Manual editing is easier to understand, and Cmake configuration too dependant on tools used.  My solution to all the complications was to copy all c and h files into one directory. I use Go with CGO and it has its own quirks and do not use cmake.

By the way, I checked out pcap library and even though it is not actively developed, it seems to be able to send using raw sockets. I belive it is installed using administrator rights, and can be used later without.

**Comment by jkvatne** [comment]
What realy put me off now, was in fact git. My local git repository as got so mingled up that I am not able to sort it out, and I am no longer able to push to my own github account.

**Comment by CapXilinx** [comment]
Yeah, CI for the Windows build would be great. Unfortunately I didn't find a free CI for any kind of Windows build (MinGW or MSVC). Warning free is also one of my goals, so I am happy for any help here.

I already rebased your commit to the current develop head, I will test it now with the conformance tester.
I invite you to contine this dicussion either on Google Groups or OpENers Gitter room https://gitter.im/EIPStackGroupOpENer/Lobby

**Comment by CapXilinx** [comment]
Rebased and merge in develop

**Comment by azoitl** [comment]
> Yeah, CI for the Windows build would be great. Unfortunately I didn't find a free CI for any kind of Windows build (MinGW or MSVC). Warning free is also one of my goals, so I am happy for any help here.
 
To not loose context I'll post one last comment here. You can use MinGW also for cross-compiling for Windows. I did first experiments in that direction for the Eclipse 4diac project. Currently only on my machine and not in CI but this could be an option for you. Especially I had even better executables with less dll dependencies from MinGW. The current 4diac FORTE 1.10.0 RC2 executable has been built that we if you want to test the result.

**Pull request closed without merging**